group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/share/ShareGroup.java

```diff
@@ -0,0 +1,299 @@
+     *                          created if it does not exist.
+     *
+     * @return A ShareGroupMember.
+     */
+    public ShareGroupMember getOrMaybeCreateMember(
+        String memberId,
+        boolean createIfNotExists
+    ) {
+        ShareGroupMember member = members.get(memberId);
+        if (member != null) return member;
+
+        if (!createIfNotExists) {
+            throw new UnknownMemberIdException(
+                String.format("Member %s is not a member of group %s.", memberId, groupId));
+        }
+
+        return new ShareGroupMember.Builder(memberId).build();
+    }
+
+    /**
+     * Updates the member.
+     *
+     * @param newMember The new share group member.
+     */
+    @Override
+    public void updateMember(ShareGroupMember newMember) {
+        if (newMember == null) {
+            throw new IllegalArgumentException("newMember cannot be null.");
+        }
+
+        ShareGroupMember oldMember = members.put(newMember.memberId(), newMember);
+        maybeUpdateSubscribedTopicNamesAndGroupSubscriptionType(oldMember, newMember);
+        maybeUpdatePartitionEpoch(oldMember, newMember);
+        maybeUpdateGroupState();
+    }
+
+    /**
+     * Remove the member from the group.
+     *
+     * @param memberId The member id to remove.
+     */
+    public void removeMember(String memberId) {
+        ShareGroupMember oldMember = members.remove(memberId);
+        maybeUpdateSubscribedTopicNamesAndGroupSubscriptionType(oldMember, null);
+        maybeRemovePartitionEpoch(oldMember);
+        maybeUpdateGroupState();
+    }
+
+    @Override
+    public void validateOffsetCommit(
+        String memberId,
+        String groupInstanceId,
+        int memberEpoch,
+        boolean isTransactional,
+        short apiVersion
+    ) {
+        throw new UnsupportedOperationException("validateOffsetCommit is not supported for Share Groups.");
+    }
+
+    @Override
+    public void validateOffsetFetch(
+        String memberId,
+        int memberEpoch,
+        long lastCommittedOffset
+    ) {
+        throw new UnsupportedOperationException("validateOffsetFetch is not supported for Share Groups.");
+    }
+
+    @Override
+    public void validateOffsetDelete() {
+        throw new UnsupportedOperationException("validateOffsetDelete is not supported for Share Groups.");
+    }
+
+    /**
+     * Validates the DeleteGroups request.
+     */
+    @Override
+    public void validateDeleteGroup() throws ApiException {
+        if (state() != ShareGroupState.EMPTY) {
+            throw Errors.NON_EMPTY_GROUP.exception();
+        }
+    }
+
+    /**
+     * Populates the list of records with tombstone(s) for deleting the group.
+     *
+     * @param records The list of records.
+     */
+    @Override
+    public void createGroupTombstoneRecords(List<CoordinatorRecord> records) {
+        records.add(CoordinatorRecordHelpers.newGroupEpochTombstoneRecord(groupId()));
```

Comment:
This should actually be created a `ShareGroupMetadata` tombstone.

Reply:
And also, the GroupMetadataManager should be writing `ShareGroupMetadata` records in `shareGroupHeartbeat`, when we get to that.

Reply:
You are right, corrected it.